### ISSUE TITLE
Using OpenGL output  10 bit color depth (GL_RGB10_A2) on a 10 bit system 

### ISSUE THREAD

**Issue by Xuno** [wang-bin/QtAV#572]
I try to play 10 bit video Sintel_480x200_yuv444p10le.mp4 on Windows 10 with 10 bit color output enabled. It plays ok, but log analysis shows 8 bit.

```Debug: "[ShaderManager] cache a new shader material type(11): gl material 16to8bit: 1, planar: 1, has alpha: 0, 2d texture: 1, 2nd plane rg: 0"```

Reference for using 10 bit color output: https://www.amd.com/Documents/10-Bit.pdf

![amd10bit](https://cloud.githubusercontent.com/assets/9219334/12524067/41c2fcb6-c163-11e5-8bd5-b5e80d5f592f.PNG)

**Comment by Xuno** [comment]
can it be tested by https://github.com/Xuno/Xuno-QtAV/commit/5be331ea6a1e69cf0388b1685acfb281d5d212e9 ?

**Comment by wang-bin** [comment]
Greate. Current code always convert to 8bit to support all devices. Fo you know how to set display depth? Does it depends on you or display?

**Comment by wang-bin** [comment]
For 10 bits yuv, we can not use GL_RGB10_A2, but GL_R16 for each plane. Actually because opengl internally use normalized float values, the output depth should equal to QSurfaceFormat.redBufferSize(). The log message `gl material 16to8bit: 1` is wrong, I should change to this `gl material 16bit-channel: 1`.

**Comment by wang-bin** [comment]
BTW, do you see my comment here https://github.com/Xuno/Xuno-QtAV/commit/af0a6f43a2ad5fc1e42b52f2acb96d8f0c69eea1

**Comment by wang-bin** [comment]
But because using `GL_R16` for 8<bpp<=15 will results in too many `ifdef` in the shader, so it's disabled by default, and use `GL_RG` instead to get the same effect. I can enable `GL_R16` and make the shader more ugly in the future(or maybe I can find a simpler way). Currently `GL_R16` can be enabled by environment var `QTAV_TEXTURE16_DEPTH=16`, but only 16bit yuv is supported. You can test it using avfilter option `format=pix_fmts=yuv444p16le`.

**Comment by wang-bin** [comment]
I will try the code in https://www.amd.com/Documents/10-Bit.pdf this week

**Comment by Xuno** [comment]
About display, yes, 10 bit display must be present and detected by AMD Catalyst (now Crimson) to have this option available.

**Comment by Xuno** [comment]
Here is a short yuv444p10le test clip

http://www.xuno.com/demo/RED_1920x822_yuv444p10le.mp4

**Comment by Xuno** [comment]
For image data greater than 8 bits isn't GL_RGB16F or GL_RGB32F as internal format required for use with GL filtering?

**Comment by wang-bin** [comment]
I'm not familiar with these formats. But I think they mean the data of 'R', 'G' and 'B' is a 16bit/32bit floating point value. Use in filters should be fine

**Comment by Xuno** [comment]
Correct, each  'R', 'G' and 'B' channel with a 16bit/32bit floating point value.  YUV source video is converted at some point to be displayed on RGB monitors in any case.

**Comment by wang-bin** [comment]
I see mpv use those formats to create fbo for high precision filtering. But it's not easy to read mpv code.

**Comment by wang-bin** [comment]
Do they only be used in filters?

**Comment by Xuno** [comment]
Possibly more, as with MPDN which supports 8- or 10- or 16-bit input and 8- or 10- or 16- bit output for maximum image quality. MPDN uses fallback depending on supported display. i.e. 16 bit video will fall back to 10 bit if 10 bit display is used.  This capability is useful for viewing RAW production video directly in different color depths without the need to make down sampled proxies.

Also, madVR uses high bitdepth processing (32bit per component).

**Comment by wang-bin** [comment]
Detect and output the display depth sounds good. I had the same idea.
Did you compared 8bit and 10bit output using MPDN? Is the difference noticible?

**Comment by Xuno** [comment]
The advantage is smooth gradients without the need for dithering:

![10bit](https://cloud.githubusercontent.com/assets/9219334/12585317/640d8e04-c400-11e5-837a-5b9ac8f6c7cd.png)

**Comment by wang-bin** [comment]
impressive

**Comment by Xuno** [comment]
I see mpv uses FBO rgba16f format as a means to display all 10 bit data.

https://github.com/mpv-player/mpv/commit/7b6e3772ab7a39fc81e7524540e69eb30c4da7ac

"So for now, run a simple conversion step through a FBO. The FBO will use
the rgba16f format normally, which gives enough bits for 10 bit, and
will at least gracefully degrade with higher depth input."

**Comment by wang-bin** [comment]
That commit breaks other things https://github.com/mpv-player/mpv/issues/2761

**Comment by Xuno** [comment]
https://ffmpeg.org/pipermail/ffmpeg-cvslog/2015-September/093840.html

ffmpeg now supports:

AVCOL_TRC_BT2020_10    = 14, ///< ITU-R BT2020 for 10 bit system
AVCOL_TRC_BT2020_12    = 15, ///< ITU-R BT2020 for 12 bit system
AVCOL_TRC_SMPTEST2084  = 16, ///< SMPTE ST 2084 for 10, 12, 14 and 16 bit systems

By using FBO with rgba16f there is room to support higher color formats depending on the user's display.

**Comment by wang-bin** [comment]
I see. If no filter, simply use a correct color matrix. Do you know how to get color matrixes? A lot of things to learn.

**Comment by Xuno** [comment]
mpv uses native client properties, so if system 10 bit color is detected color matrix, etc. is from system. 

MPV_FORMAT_NODE_MAP
    "pixelformat"       MPV_FORMAT_STRING
    "w"                 MPV_FORMAT_INT64
    "h"                 MPV_FORMAT_INT64
    "dw"                MPV_FORMAT_INT64
    "dh"                MPV_FORMAT_INT64
    "aspect"            MPV_FORMAT_DOUBLE
    "par"               MPV_FORMAT_DOUBLE
    "colormatrix"       MPV_FORMAT_STRING
    "colorlevels"       MPV_FORMAT_STRING
    "primaries"         MPV_FORMAT_STRING
    "chroma-location"   MPV_FORMAT_STRING
    "rotate"            MPV_FORMAT_INT64
    "stereo-in"         MPV_FORMAT_STRING

**Comment by wang-bin** [comment]
I find that I can't create 10bit window mentioned in the pdf. I also tried to change QSurfaceFormat, for example change opengl version, but it didn't work. Can you try 10bit QSurfaceFormat? I think it's the only way to support 10 bit in qt.

**Comment by Xuno** [comment]
We get the same results.  We will do more testing.

**Comment by wang-bin** [comment]
Maybe it's a qt bug. Does mpv support 10bit window?
I also change gl code to use r16 instead of rg texture to render 10bit videos

**Comment by Xuno** [comment]
madVR + MPC-HC in **fullscreen exclusive mode** I can output with 10 bit.
